internal/distro/fedora/distro.go

```diff
@@ -196,49 +211,73 @@ func (t *imageType) PartitionType() string {
 	return ""
 }
 
-func (t *imageType) Packages(bp blueprint.Blueprint) ([]string, []string) {
-	packages := append(t.packages, bp.GetPackages()...)
-	timezone, _ := bp.Customizations.GetTimezoneSettings()
-	if timezone != nil {
-		packages = append(packages, "chrony")
+func (t *imageType) getPackages(name string) rpmmd.PackageSet {
+	getter := t.packageSets[name]
+	if getter == nil {
+		return rpmmd.PackageSet{}
+	}
+
+	return getter(t)
+}
+
+func (t *imageType) PackageSets(bp blueprint.Blueprint) map[string]rpmmd.PackageSet {
+	// merge package sets that appear in the image type with the package sets
+	// of the same name from the distro and arch
+	mergedSets := make(map[string]rpmmd.PackageSet)
+
+	imageSets := t.packageSets
+
+	for name := range imageSets {
+		mergedSets[name] = t.getPackages(name)
 	}
+
+	if _, hasPackages := imageSets[osPkgsKey]; !hasPackages {
+		// should this be possible??
+		mergedSets[osPkgsKey] = rpmmd.PackageSet{}
+	}
+
+	// every image type must define a 'build' package set
+	if _, hasBuild := imageSets[buildPkgsKey]; !hasBuild {
+		panic(fmt.Sprintf("'%s' image type has no '%s' package set defined", t.name, buildPkgsKey))
+	}
+
 	if t.bootable {
-		packages = append(packages, t.arch.bootloaderPackages...)
+		switch t.arch.Name() {
+		case distro.X86_64ArchName:
+			mergedSets[osPkgsKey] = mergedSets[osPkgsKey].Append(rpmmd.PackageSet{Include: t.arch.bootloaderPackages})
+		case distro.Aarch64ArchName:
+			mergedSets[osPkgsKey] = mergedSets[osPkgsKey].Append(rpmmd.PackageSet{Include: t.arch.bootloaderPackages})
+		}
+
 	}
+	// blueprint packages
+	bpPackages := bp.GetPackages()
+	timezone, _ := bp.Customizations.GetTimezoneSettings()
+	if timezone != nil {
+		bpPackages = append(bpPackages, "chrony")
+	}
+
+	// depsolve bp packages separately
+	// bp packages aren't restricted by exclude lists
+	mergedSets[blueprintPkgsKey] = rpmmd.PackageSet{Include: bpPackages}
 
 	// copy the list of excluded packages from the image type
 	// and subtract any packages found in the blueprint (this
 	// will not handle the issue with dependencies present in
 	// the list of excluded packages, but it will create a
 	// possibility of a workaround at least)
-	excludedPackages := append([]string(nil), t.excludedPackages...)
-	for _, pkg := range bp.GetPackages() {
+	excludedPackages := mergedSets[osPkgsKey].Exclude
+	for _, pkg := range mergedSets[osPkgsKey].Append(mergedSets[blueprintPkgsKey]).Include {
 		// removePackage is fine if the package doesn't exist
 		excludedPackages = removePackage(excludedPackages, pkg)
 	}
+	mergedSets[osPkgsKey] = rpmmd.PackageSet{Include: mergedSets[osPkgsKey].Include, Exclude: excludedPackages}
 
-	return packages, excludedPackages
-}
+	// Add kernel package if none defined
+	kc := bp.Customizations.GetKernel()
+	mergedSets[osPkgsKey] = mergedSets[osPkgsKey].Append(rpmmd.PackageSet{Include: []string{kc.Name}})
```

Comment:
I have to apologize, but after a discussion within our team, we agreed that the preferred approach for now is to use the same approach as used in RHEL-8.6/9.0 distro definitions. This means introducing #2386 , but we think that it is acceptable. So this means reverting my previous suggestion and depsolving the "os" and "blueprint" packages separately. Sorry for the noise 🙄